File: design-pattern/state-pattern/state.ts

interface VendingState {
    insertCoin(machine: VendinMachine, coin: number): VendingState;
    selectItem(machine: VendinMachine): VendingState;
    dispensItem(machine: VendinMachine): VendingState;
    returnCoin(machine: VendinMachine): VendingState;
    refillQuant(machine: VendinMachine, quant: number): VendingState;
    stateStatus(): string;
};

class VendinMachine {
    private currState: VendingState;
    private noCoinState: VendingState;
    private hasCoinState: VendingState;
    private dispenseState: VendingState;
    private soldoutState: VendingState;
    private quatity: number = 0;
    private price: number = 0;
    private currentCoin: number = 0;

    constructor(quant: number, price: number) {
        this.quatity = quant
        this.price = price;
        this.currentCoin = 0;


        this.noCoinState = new NoCoinState();
        this.hasCoinState = new HasCoinState();
        this.dispenseState = new DispenseState();
        this.soldoutState = new SoldoutState();

        if (this.quatity > 0) {
            this.currState = new NoCoinState();
        } else {
            this.currState = new SoldoutState();
        }
    };

    insertCoin(coin: number): void {
        this.currState = this.currState.insertCoin(this, coin);
    }
    selectItem(): void {
        this.currState = this.currState.selectItem(this);
    }

    dispensItem(): void {
        this.currState = this.currState.dispensItem(this);
    }
    returnCoin(): void {
        this.currState = this.currState.returnCoin(this);
    }
    refillQuant(quant: number): void {
        this.currState = this.currState.refillQuant(this, quant);
    };

    getNoCoinState() { return this.noCoinState; }
    getHasCoinState() { return this.hasCoinState; }
    getDispenseState() { return this.dispenseState; }
    getSoldOutState() { return this.soldoutState; }

    getItemCount() { return this.quatity; }
    decrementItem() { this.quatity--; }
    incrementItem(qty: number) { this.quatity += qty; }

    getBalance() { return this.currentCoin; }
    setBalance(amount: number) { this.currentCoin = amount; }
    addBalance(amount: number) { this.currentCoin += amount; }

    getPrice() { return this.price; }


    stateStatus(): void {
        console.log("\n--- VENDING MACHINE STATUS ---");
        console.log("Items remaining:", this.quatity);
        console.log("Balance:", this.currentCoin);
        console.log("State:", this.currState.stateStatus());
    }
};

class NoCoinState implements VendingState {
    insertCoin(machine: VendinMachine, coin: number): VendingState {
        machine.setBalance(coin);
        console.log(`Coin inserted: Rs ${coin}`);
        return machine.getHasCoinState();
    }
    selectItem(machine: VendinMachine): VendingState {
        console.log("please insert the coin");
        return machine.getNoCoinState();
    }
    dispensItem(machine: VendinMachine): VendingState {
        console.log("please insert the coin");
        return machine.getNoCoinState();
    }
    returnCoin(machine: VendinMachine): VendingState {
        console.log("please insert the coin");
        return machine.getNoCoinState();
    }
    refillQuant(machine: VendinMachine, quant: number): VendingState {
        console.log("item refilling");
        machine.incrementItem(quant);
        return machine.getNoCoinState();
    }
    stateStatus(): string {
        return "NO_COIN_STATE"
    }
};

class HasCoinState implements VendingState {
    insertCoin(machine: VendinMachine, coin: number): VendingState {
        machine.addBalance(coin);
        console.log(`Added coin. Balance: Rs ${machine.getBalance()}`);
        return this;
    }

    selectItem(machine: VendinMachine): VendingState {
        if (machine.getItemCount() === 0) {
            console.log("Machine is sold out!");
            return machine.getSoldOutState();
        }

        if (machine.getBalance() >= machine.getPrice()) {
            console.log("Item selected. Preparing to dispense...");

            const change = machine.getBalance() - machine.getPrice();
            if (change > 0) {
                console.log(`Returning change: Rs ${change}`);
            }

            machine.setBalance(0);
            return machine.getDispenseState();
        } else {
            const needed = machine.getPrice() - machine.getBalance();
            console.log(`Insufficient balance. Need Rs ${needed} more.`);
            return this;
        }
    }

    dispensItem(): VendingState {
        console.log("Please select item first!");
        return this;
    }

    returnCoin(machine: VendinMachine): VendingState {
        console.log(`Returning coin: Rs ${machine.getBalance()}`);
        machine.setBalance(0);
        return machine.getNoCoinState();
    }

    refillQuant(): VendingState {
        console.log("Cannot refill while transaction is in progress.");
        return this;
    }

    stateStatus(): string {
        return "HAS_COIN";
    }
}

// DispenseState
class DispenseState implements VendingState {
    insertCoin(_: VendinMachine, coin: number): VendingState {
        console.log(`Please wait. Dispensing... Coin returned: Rs ${coin}`);
        return this;
    }

    selectItem(): VendingState {
        console.log("Already dispensing item...");
        return this;
    }

    dispensItem(machine: VendinMachine): VendingState {
        console.log("Item dispensed!");
        machine.decrementItem();

        if (machine.getItemCount() > 0) {
            return machine.getNoCoinState();
        } else {
            console.log("Machine is now SOLD OUT!");
            return machine.getSoldOutState();
        }
    }

    returnCoin(): VendingState {
        console.log("Cannot return coin while dispensing!");
        return this;
    }

    refillQuant(): VendingState {
        console.log("Cannot refill during dispensing.");
        return this;
    }

    stateStatus(): string {
        return "DISPENSING";
    }
}

// SoldOutState
class SoldoutState implements VendingState {
    insertCoin(_: VendinMachine, coin: number): VendingState {
        console.log(`Machine sold out. Returning Rs ${coin}`);
        return this;
    }

    selectItem(): VendingState {
        console.log("Machine is sold out!");
        return this;
    }

    dispensItem(): VendingState {
        console.log("Machine is sold out!");
        return this;
    }

    returnCoin(): VendingState {
        console.log("No coin inserted.");
        return this;
    }

    refillQuant(machine: VendinMachine, quantity: number): VendingState {
        machine.incrementItem(quantity);
        console.log("Machine refilled.");
        return machine.getNoCoinState();
    }

    stateStatus(): string {
        return "SOLD_OUT";
    }
}

// ---------------- MAIN ----------------

const machine = new VendinMachine(2, 20);

console.log("=== VENDING MACHINE START ===");

machine.stateStatus();

console.log("\n1. Select without coin:");
machine.selectItem();
machine.stateStatus();

console.log("\n2. Insert coin:");
machine.insertCoin(10);
machine.stateStatus();

console.log("\n3. Select with insufficient funds:");
machine.selectItem();
machine.stateStatus();

console.log("\n4. Add more coins:");
machine.insertCoin(10);
machine.stateStatus();

console.log("\n5. Select item:");
machine.selectItem();
machine.stateStatus();

console.log("\n6. Dispense item:");
machine.dispensItem();
machine.stateStatus();

console.log("\n7. Buy last item:");
machine.insertCoin(20);
machine.selectItem();
machine.dispensItem();
machine.stateStatus();

console.log("\n8. Try using sold-out machine:");
machine.insertCoin(5);

console.log("\n9. Refill machine:");
machine.refillQuant(2);
machine.stateStatus();

File: design-pattern/state-pattern/state.cpp
#include <iostream>
#include <string>

using namespace std;

class VendingMachine;
class VendingState;

// states
class NocoinState;
class HascoinState;
class DispanceState;
class SoldoutState;

// abstract class
class VendingState
{
    virtual VendingState *insertCoin(VendingMachine *machine, int coin) = 0;
    virtual VendingState *selectItem(VendingMachine *machine) = 0;
    virtual VendingState *dispenseItem(VendingMachine *machine) = 0;
    virtual VendingState *ReturnCoin(VendingMachine *machine) = 0;
    virtual VendingState *Refil(VendingMachine *machine, int quant) = 0;
    virtual string printStatus() = 0;
};

class VendingMachine
{
private:
    VendingState *currState;
    int quantity;
    int price;
    int insertedCoin;

    // state object
    VendingState *noCoinState;
    VendingState *hasCoinState;
    VendingState *dispenseState;
    VendingState *soldOutState;

public:
    VendingMachine(int quantity, int price);

    void insertCoin(int coin);
    void selectItem();
    void dispenseItem();
    void ReturnCoin();
    void Refil(int quant);
    void printStatus();

    // curd operation
    VendingState *getNoCoinState()
    {
        return NocoinState;
    };

    VendingState *getHasCoinState()
    {
        return HascoinState;
    };

    VendingState *getDispenseState()
    {
        return DispanceState;
    };

    VendingState *getSoldOutState()
    {
        return SoldoutState;
    };

    int getQuantity()
    {
        this->quantity;
    };

    int getInstertedCoin()
    {
        this->insertedCoin;
    };

    void setInsertCoin(int coin)
    {
        this->insertedCoin += coin;
    };

    void addCoin(int coin)
    {
        this->insertedCoin = coin
    }

    void
    insertQuantity(int quant)
    {
        this->quantity += quant;
    };

    void removeQuantity()
    {
        this->quantity--;
    }

    int getPrice()
    {
        this->price;
    }

    void setPrice(int price)
    {
        this->price = price;
    }
};

class NocoinState : public VendingState
{
    VendingState *insertCoin(VendingMachine *machine, int coin) override
    {
        machine->addCoin(coin);
        cout << "the coin is inserted" << coin << endl;
        return machine->getHasCoinState();
    };

    VendingMachine *selectItem(VendingMachine *machine) override
    {
        cout << "please insert the coin" << endl;
        return machine->getNoCoinState();
    };

    VendingMachine *dispenseItem(VendingMachine *machine) override
    {
        cout << "please insert the coin" << endl;
        return machine->getNoCoinState();
    };

    VendingMachine *ReturnCoin(VendingMachine *machine) override
    {
        cout << "please insert the coin" << endl;
        return machine->getNoCoinState();
    };

    VendingMachine *Refil(VendingMachine *machine, int quant) override
    {
        cout << "item is refilling" << endl;
        machine->insertQuantity(quant);
        return machine->getNoCoinState();
    };

    string printStatus() override
    {
        return "NO_COIN";
    }
};

class HascoinState : public VendingState
{
    VendingState *insertCoin(VendingMachine *machine, int coin) override
    {
        machine->setInsertCoin(coin);
        cout << "the coin is inserted" << coin << endl;
        return machine->getHasCoinState();
    };

    VendingMachine *selectItem(VendingMachine *machine) override
    {
        if (machine->getInstertedCoin() >= machine->getPrice())
        {
            cout << "item is dispensing"

                int amount = machine->getInstertedCoin() - machine->getPrice();

            if amount
                > 0
                {
                    cout << "change is returning with amount : " << amount << endl;
                }
            machine->setInsertedCoin(0);

            return machine->getDispenseState();
        }
        else
        {
            int needed = machine->getPrice() - machine->getInsertedCoin();
            cout << "Insufficient funds. Need Rs " << needed << " more." << endl;
            return machine->getHasCoinState(); // Stay in same state
        }
    };

    VendingMachine *dispenseItem(VendingMachine *machine) override
    {
        cout << "please select the item to dispense" << endl;
        return machine->getHasCoinState();
    };

    VendingMachine *ReturnCoin(VendingMachine *machine) override
    {
        cout << "amount inserted is : " << machine->getInstertedCoin();
        << endl;
        cout << machine->addCoin(0);
        return machine->getNoCoinState();
    };

    VendingMachine *Refil(VendingMachine *machine, int quant) override
    {
        cout << "can't refill in this state" << endl;
        return machine->getHasCoinState();
    };

    string printStatus() override
    {
        return "NO_COIN";
    }
};

class DispanceState : public VendingState
{
    VendingState *insertCoin(VendingMachine *machine, int coin) override
    {
        cout << "sorry wait for ieam to dispense" << coin << endl;
        return machine->getDispenseState();
    };

    VendingMachine *selectItem(VendingMachine *machine) override
    {
        cout << "wait you already select the item" << endl;

        return machine->getDispenseState();
    };

    VendingMachine *dispenseItem(VendingMachine *machine) override
    {
        cout << "item is dispensed" << endl;

        machine->removeQuantity();

        if (machine->getQuantity() > 0)
        {
            return machine->getNoCoinState();
        }
        else
        {
            cout << "Sorry Item is OutOfStock" << endl;

            return machine->getSoldOutState();
        }
    };

    VendingMachine *ReturnCoin(VendingMachine *machine) override
    {
        cout << "you can't get coin this state : ";
        << endl;
        return machine->getDispenseState();
    };

    VendingMachine *Refil(VendingMachine *machine, int quant) override
    {
        cout << "can't refill in this state" << endl;
        return machine->getDispenseState();
    };

    string printStatus() override
    {
        return "DISPENSE_ITEM";
    }
};

class SoldOutState : public VendingState
{
    VendingState *insertCoin(VendingMachine *machine, int coin) override
    {
        cout << "the item is sold out" << coin << endl;
        return machine->getSoldOutState();
    };

    VendingMachine *selectItem(VendingMachine *machine) override
    {
        cout << "the item is sold out" << endl;

        return machine->getSoldOutState();
    };

    VendingMachine *dispenseItem(VendingMachine *machine) override
    {
        cout << "ithe item is sold out" << endl;
        return machine->getSoldOutState();
    };

    VendingMachine *ReturnCoin(VendingMachine *machine) override
    {
        cout << " sorry the iteam is out of stock ";
        << endl;
        return machine->getSoldOutState();
    };

    VendingMachine *Refil(VendingMachine *machine, int quant) override
    {
        machine->insertQuantity(quant);
        return machine->getNoCoinState();
    };

    string printStatus() override
    {
        return "SOLD_OUT";
    };
};

VendingMachine::VendingMachine(int item, int price)
{
    this->quantity = item;
    this->price = price;
    this->insertedCoin = 0;

    nocoinState = new NocoinState();
    hasCoinState = new HascoinState();
    dispenseState = new DispanceState();
    soldOutState = new SoldOutState();

    if (item > 0)
    {
        currState = noCoinState;
    }
    else
    {
        currState = soldOutState;
    }
};

void VendingMachine::insertCoin(int coin)
{
    currState = currState->insertCoin(this, coin);
}

void VendingMachine::selectItem()
{
    currentState = currentState->selectItem(this);
}

void VendingMachine::dispenseItem()
{
    currentState = currentState->dispenseItem(this);
}

void VendingMachine::ReturnCoin()
{
    currentState = currentState->ReturnCoin(this);
}

void VendingMachine::refill(int quantity)
{
    currentState = currentState->refill(this, quantity);
}

void VendingMachine::printStatus()
{
    cout << "\n--- Vending Machine Status ---" << endl;
    cout << "Items remaining: " << itemCount << endl;
    cout << "Inserted coin: Rs " << insertedCoins << endl;
    cout << "Current state: " << currentState->getStateName() << endl
         << endl;
}

int main(void)
{
    int itemCount = 5;
    int itemPrice = 10;

    VendingMachine machine(itemCount, itemPrice);
    machine.printStatus();

    cout << "1. Trying to select item without coin:" << endl;
    machine.selectItem(); // Should ask for coin, no state change
    machine.printStatus();

    cout << "2. Inserting coin:" << endl;
    machine.insertCoin(10); // State changes to HAS_COIN
    machine.printStatus();

    cout << "3. Selecting item with insufficient funds:" << endl;
    machine.selectItem(); // Insufficient funds, stays in HAS_COIN
    machine.printStatus();

    cout << "4. Adding more coins:" << endl;
    machine.insertCoin(10); // Add more money, stays in HAS_COIN
    machine.printStatus();

    cout << "5. Selecting item Now" << endl;
    machine.selectItem(); // State changes to SOLD
    machine.printStatus();

    cout << "6. Dispensing item:" << endl;
    machine.dispenseItem(); // State changes to NO_COIN (items remaining)
    machine.printStatus();

    cout << "7. Buying last item:" << endl;
    machine.insertCoin(20); // State changes to HAS_COIN
    machine.selectItem();   // State changes to SOLD
    machine.dispenseItem(); // State changes to SOLD_OUT (no items left)
    machine.printStatus();

    cout << "8. Trying to use sold out machine:" << endl;
    machine.insertCoin(5); // Coin returned, stays in SOLD_OUT

    cout << "9. Trying to use sold out machine:" << endl;
    machine.refill(2);
    machine.printStatus(); // State changes NO_COIN

    return 0;
};